Realm/Tests/RealmTests.mm

```diff
@@ -1455,6 +1455,438 @@ - (void)testActiveVersionLimit {
+    // notification shouldnt have fired
+    XCTAssertTrue(notificationFired);
+    [token invalidate];
+}
+
+- (void)testBeginAsyncTransactionInAsyncTransaction {
+    RLMRealm *realm = [self realmWithTestPath];
+    XCTestExpectation *transaction1 = [self expectationWithDescription:@"async transaction 1 complete"];
+    XCTestExpectation *transaction2 = [self expectationWithDescription:@"async transaction 2 complete"];
+
+    [realm beginAsyncWriteTransaction:^{
+        NSError *error = nil;
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+        
+        [realm beginAsyncWriteTransaction:^{
+            [StringObject createInRealm:realm withValue:@[@"string"]];
+            
+            [realm commitAsyncWriteTransaction:^(NSError *error) {
+                XCTAssertEqual(0, [StringObject allObjects].count);
+                [transaction1 fulfill];
+                XCTAssertNil(error);
+            }];
+        }];
+
+        [realm commitAsyncWriteTransaction:^(NSError *error) {
+            XCTAssertEqual(0, [StringObject allObjects].count);
+            [transaction2 fulfill];
+            XCTAssertNil(error);
+        }];
+    }];
+    
+    [self waitForExpectationsWithTimeout:3.0 handler:nil];
+    XCTAssertEqual(0, [StringObject allObjects].count);
+}
+
+- (void)testAsyncTransactionFromSyncTransaction {
+    RLMRealm *realm = [self realmWithTestPath];
+    XCTestExpectation *transaction1 = [self expectationWithDescription:@"async transaction 1 complete"];
+
+    [realm beginWriteTransaction];
+    [StringObject createInRealm:realm withValue:@[@"string"]];
+
+    [realm beginAsyncWriteTransaction:^{
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+        NSError *error = nil;
+
+        [realm commitAsyncWriteTransaction:^(NSError *error) {
+            XCTAssertEqual(0, [StringObject allObjects].count);
+            [transaction1 fulfill];
+            XCTAssertNil(error);
+        }];
+    }];
+    [realm commitWriteTransaction];
+    
+    [self waitForExpectationsWithTimeout:3.0 handler:nil];
+    XCTAssertEqual(0, [StringObject allObjects].count);
+}
+
+- (void)testAsyncTransactionCancel {
+    RLMRealm *realm = [RLMRealm defaultRealm];
+    auto expectation = [self expectationWithDescription:@""];
+    expectation.expectedFulfillmentCount = 4;
+
+    [realm beginAsyncWriteTransaction:^{
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+        [expectation fulfill];
+    }];
+
+    [realm beginAsyncWriteTransaction:^{
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+        [expectation fulfill];
+        [realm commitAsyncWriteTransaction];
+    }];
+
+    RLMAsyncTransactionId asyncTransactionIdA = [realm beginAsyncWriteTransaction:^{
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+        [expectation fulfill];
+        [realm cancelAsyncTransaction:asyncTransactionIdA];
```

Comment:
This doesn't actually do anything; the transaction is just being implicitly cancelled due to to commit.